presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1320,87 +1621,110 @@ private void uploadObject()
+                CompletedFileUpload completedUpload = fileUpload.completionFuture().join();
 
-                upload.waitForCompletion();
                 STATS.uploadSuccessful();
-                log.debug("Completed upload for host: %s, key: %s", host, key);
-            }
-            catch (AmazonClientException e) {
-                STATS.uploadFailed();
-                throw new IOException(e);
+                log.debug("Completed upload for host: %s, key: %s, ETag: %s",
+                        host, key, completedUpload.response().eTag());
             }
-            catch (InterruptedException e) {
+            catch (Exception e) {
                 STATS.uploadFailed();
-                Thread.currentThread().interrupt();
-                throw new InterruptedIOException();
+                if (e.getCause() instanceof InterruptedException) {
+                    Thread.currentThread().interrupt();
+                    throw new InterruptedIOException();
+                }
+                throw new IOException("Upload failed for key: " + key, e);
             }
         }
 
-        private ProgressListener createProgressListener(Transfer transfer)
+        private TransferListener createTransferListener()
         {
-            return new ProgressListener()
-            {
-                private ProgressEventType previousType;
-                private double previousTransferred;
+            return new TransferListener() {
+                private long previousBytesTransferred;
 
                 @Override
-                public synchronized void progressChanged(ProgressEvent progressEvent)
+                public void transferInitiated(Context.TransferInitiated context)
                 {
-                    ProgressEventType eventType = progressEvent.getEventType();
-                    if (previousType != eventType) {
-                        log.debug("Upload progress event (%s/%s): %s", host, key, eventType);
-                        previousType = eventType;
-                    }
+                    log.debug("Upload initiated for %s/%s", host, key);
+                }
 
-                    double transferred = transfer.getProgress().getPercentTransferred();
-                    if (transferred >= (previousTransferred + 10.0)) {
-                        log.debug("Upload percentage (%s/%s): %.0f%%", host, key, transferred);
-                        previousTransferred = transferred;
+                @Override
+                public void bytesTransferred(Context.BytesTransferred context)
+                {
+                    long currentBytes = context.progressSnapshot().transferredBytes();
+                    if (currentBytes - previousBytesTransferred >= 10 * 1024 * 1024) {
+                        log.debug("Upload progress (%s/%s): %d bytes transferred",
+                                host, key, currentBytes);
+                        previousBytesTransferred = currentBytes;
```

Comment:
Since we were using `transfer.getProgress().getPercentTransferred()`, maybe `context.progressSnapshot().ratioTransferred()` might be more suitable in the v2 implementation.